src/components/App.js

```diff
@@ -2,18 +2,103 @@ import React from 'react';
 import '../App.css';
 import Display from './Display';
 import ButtonPanel from './ButtonPanel';
+// eslint-disable-next-line
+import calculate from '../logic/calculate';
 
 class App extends React.Component {
   constructor(props) {
     super(props);
-    this.state = {};
+    this.handleClick = this.handleClick.bind(this);
+    this.state = {
+      total: null,
+      next: null,
+      operation: null,
+    };
+  }
+
+  handleClick(buttonName) {
+    const { total, next, operation } = this.state;
+    if (buttonName === 'AC') {
+      this.setState({
+        total: null,
+        next: null,
+        operation: null,
+      });
+    } else if (buttonName === '+/-') {
+      const newState = calculate(this.state, buttonName);
+      const newTotal = newState.total;
+      const newNext = newState.next;
+      this.setState({ total: newTotal, next: newNext });
+    } else if (Number(buttonName) || Number(buttonName) === 0) {
+      if (!operation) {
+        if (total) {
+          if (total === '0') {
+            this.setState({
+              total: buttonName,
+            });
+          } else {
+            this.setState({
+              total: total.concat(buttonName),
+            });
+          }
+        } else {
+          this.setState({
+            total: buttonName,
+          });
+        }
+      } else if (next) {
+        this.setState({
+          next: next.concat(buttonName),
+        });
+      } else {
+        this.setState({
+          next: buttonName,
+        });
+      }
+    } else if (buttonName === '.') {
+      if (!operation) {
+        if (!(/\D/).test(total)) {
+          this.setState({
+            total: total.concat('.'),
+          });
+        }
+      } else if (!(/\D/).test(next)) {
+        this.setState({
+          next: next.concat('.'),
+        });
+      }
+    } else if (buttonName === '=') {
+      if (total && next && operation) {
+        const result = calculate(this.state, buttonName);
+        const newTotal = result.total;
+        this.setState({
+          total: newTotal,
+          next: null,
+          operation: null,
+        });
+      }
+    } else {
+      this.setState({
+        operation: buttonName,
+      });
+    }
```

Comment:
there's a reason we have a LOGIC folder, this logic here should be inside `calculate.js`, thats why you imported calculate in this Component, so that you can use it inside this `handleClick()` function. So now transfer the code there, and just call it here

Reply:
Thanks, I will correct that!

Reply:
Hello Shay, can you explain more about putting this logic in the calculate? Because I am setting the state based on the conditions and the values from calling the calculate (in the handleClick function). I used the calculate to only calculate values and return them. Thanks

Reply:
I know that the values of the state are useful here :
But you have to implement this logic inside `calculate.js` , then you call the function at this level and pass in the states as parameter, just like you done in line `18`. 

That's  reason `calculate` was asked to be imported here as you did in line 6 (on milestone 4)

Also this `calculate()` function that is supposed to be inside `calculate.js` returns  new values of ` { total, next, operation }` after calculation, which you will use at this level insode `</App>` to set new state values.

----------------------
At the send, you free up this function, to look some like 
```
handleClick = (buttonName) => {
    const { total, next, operation } = calculate(this.state, buttonName);
    this.setState({
      total: total,
      next: next,
      operation: operation
    })
  }
```

or more shorter to something like :
``` 

  handleClick = (buttonName) => {
  this.setState(calculate(this.state, buttonName));
  }


```